### PR TITLE
1000 typeahead menu button

### PR DESCRIPTION
**Pull request by austingardner**
## Description 
This PR adds the enhanced multicharacter typeahead functionality that is added to makeup to the menu. 
I also added tests to the menu button to assure that this functionality is working. 
Added makeup-typeahead to deps. 
Updated makeup-floating-label to most recent version. 

## References
https://github.com/eBay/ebayui-core/issues/1000

## Screenshots
Already demo'd -> can add them if needed

### CODE REVIEW

File: src/components/ebay-menu/component.js [hunk 1]
```diff
@@ -2,9 +2,11 @@ const assign = require('core-js-pure/features/object/assign');
 const findIndex = require('core-js-pure/features/array/find-index');
 const scrollKeyPreventer = require('makeup-prevent-scroll-keys');
 const rovingTabindex = require('makeup-roving-tabindex');
+const typeahead = require('makeup-typeahead');
 const eventUtils = require('../../common/event-utils');
 const menuUtils = require('../../common/menu-utils');
-const NodeListUtils = require('../../common/nodelist-utils');
+
+const TIMEOUT_LENGTH = 1300;
```
Comment (ianmcburnie) [inline src/components/ebay-menu/component.js:9]: Maybe just rename this variable so that its purpose is clear. Maybe `TYPEAHEAD_TIMEOUT_LENGTH`.

File: src/components/ebay-menu/component.js [hunk 2]
```diff
@@ -48,7 +50,7 @@ module.exports = assign({}, menuUtils, {
     },
 
     handleItemKeypress({ key }) {
-        const itemIndex = NodeListUtils.findNodeWithFirstChar(this.getEl('menu').children, key);
+        const itemIndex = typeahead(this.getEl('menu').children, key, TIMEOUT_LENGTH);
```
Comment (ianmcburnie) [inline src/components/ebay-menu/component.js:53]: Nice clean interface. I like it.

Comment (ianmcburnie) [reply src/components/ebay-menu/component.js:53]: Just thought though, do you cover for edge cases like missing menu element or zero children?

Comment (austingardner) [reply src/components/ebay-menu/component.js:53]: yes -> added tests in the makeup repo for those cases

**Review by ianmcburnie — COMMENTED** [review]
I think it would be nice to have a new story/example that better demonstrates the new typeahead.

**Review by ianmcburnie — COMMENTED** [review]
It seems to be doing a kind of fuzzy-match now. In the new example, when I press 'i' it goes to the first entry that has an 'i' anywhere in it, and not at the start. Is this intentional?

**Comment by austingardner** [comment]
Yes, the fuzzy match is intentional. If there is no match at the start of the string, then the typeahead will look anywhere.

**Comment by ianmcburnie** [comment]
Okay, wasn't totally expecting it, but we can rollout with that and can always make it configurable to offer a more linear option if necessary.

File: src/components/ebay-menu-button/examples/18-typeahead/template.marko [hunk 3]
```diff
@@ -0,0 +1,53 @@
+<style>
```
Comment (ianmcburnie) [inline src/components/ebay-menu-button/examples/18-typeahead/template.marko:1]: You can removed this style section from this example as it's not used.

Comment (austingardner) [reply src/components/ebay-menu-button/examples/18-typeahead/template.marko]: done

**Review by ianmcburnie — APPROVED** [review]

File: src/components/ebay-menu/component.js [hunk 4]
```diff
@@ -2,9 +2,11 @@ const assign = require('core-js-pure/features/object/assign');
 const findIndex = require('core-js-pure/features/array/find-index');
 const scrollKeyPreventer = require('makeup-prevent-scroll-keys');
 const rovingTabindex = require('makeup-roving-tabindex');
+const typeahead = require('makeup-typeahead');
 const eventUtils = require('../../common/event-utils');
 const menuUtils = require('../../common/menu-utils');
-const NodeListUtils = require('../../common/nodelist-utils');
+
+const TYPEAHEAD_TIMEOUT_LENGTH = 1300;
```
Comment (agliga) [inline src/components/ebay-menu/component.js:9]: @ianmcburnie Would we want to allow this value to be overridden by teams?
If so we can do this where we read the value: `this.input.typeaheadTimeoutLength || TYPEAHEAD_TIMEOUT_LENGTH;`

File: src/components/ebay-menu/component.js [hunk 5]
```diff
@@ -118,12 +120,19 @@ module.exports = assign({}, menuUtils, {
         });
 
         scrollKeyPreventer.add(this.contentEl);
+
+        const { getIndex, destroy } = typeahead();
+        this.getIndex = getIndex;
```
Comment (agliga) [inline src/components/ebay-menu/component.js:125]: Can we call this something like: `getTypeaheadIndex` or something else? `getIndex` seems pretty generic in this case and might be a confusing name.

File: src/components/ebay-menu/test/mock/index.js [hunk 6]
```diff
@@ -30,3 +30,19 @@ exports.Fake_2Items = assign({}, exports.Basic_2Items, {
         renderBody: createRenderBody(`Item text ${i}`)
     }))
 });
+
+function getCountry(num) {
```
Comment (agliga) [inline src/components/ebay-menu/test/mock/index.js:34]: What's the point of having this as a function?
You can probably use an array instead with various countries. 
Like: 
const countries = ['Albania', 'Alcania', 'Something else'];

`item ${i} ${countries[i]}`;

File: src/components/ebay-menu/test/test.browser.js [hunk 7]
```diff
@@ -11,6 +11,29 @@ afterEach(cleanup);
 /** @type import("@marko/testing-library").RenderResult */
 let component;
 
+describe('typeahead functionality', () => {
```
Comment (agliga) [inline src/components/ebay-menu/test/test.browser.js:14]: Do we want a test to check the timeout as well, like:
Type a
Wait for 2 seconds
Type l
Should focus on item starting with l

Or is this covered in the makeup portion?

Also do we want to check the functionality if nothing matches? Like if someone types in "a", "l" "x" that nothing is focused or something, or the last one is still focused?

**Review by agliga — COMMENTED** [review]
Couple of comments

Overall looks really good. I think just needs some minor cleanup (and some of these are pretty nitpicky as well)

File: src/components/ebay-menu-button/examples/18-typeahead/template.marko [hunk 8]
```diff
@@ -0,0 +1,41 @@
+<h3>
+    Click the button, then type abcdefghijk to see focus change.
```
[the remark this answers is not in the document]
Comment (austingardner) [reply src/components/ebay-menu-button/examples/18-typeahead/template.marko:2]: I guess what I was going for was that they could type and sequentially go down the list. typing ab then abc would move the focus from ab to abc. But yeah if you don't like it we could change it.

[the remark this answers is not in the document]
Comment (agliga) [reply src/components/ebay-menu-button/examples/18-typeahead/template.marko:2]: That's fine, lets just think of a better example. 
Maybe even Country, State, city, like
`United States`
`United States, California`
`United States, California, Los Angeles`
`United States, California, San Francisco`
`United States, California, San Jose`
Etc
Just something that came off the top of my head. We can do something else, maybe countries or such.

Ideally this would cover users usecases since the abc isn't a real usecase, but we can think of things that can coincide.

[the remark this answers is not in the document]
Comment (ianmcburnie) [reply src/components/ebay-menu-button/examples/18-typeahead/template.marko:2]: Yeah, I think we should change it for something more real world

File: src/components/ebay-menu/component.js (hunk 4 above)
Comment (ianmcburnie) [reply src/components/ebay-menu/component.js:9]: I don't think we'd want them to override.

File: src/components/ebay-menu/component.js (hunk 5 above)
Comment (ianmcburnie) [reply src/components/ebay-menu/component.js:125]: Agreed.

File: src/components/ebay-menu/test/test.browser.js (hunk 7 above)
Comment (ianmcburnie) [reply src/components/ebay-menu/test/test.browser.js:14]: I think the nothing matches would be a good test case.

Comment (austingardner) [reply src/components/ebay-menu/test/test.browser.js:14]: done

File: src/components/ebay-menu/component.js [hunk 9]
```diff
@@ -2,9 +2,11 @@ const assign = require('core-js-pure/features/object/assign');
 const findIndex = require('core-js-pure/features/array/find-index');
 const scrollKeyPreventer = require('makeup-prevent-scroll-keys');
 const rovingTabindex = require('makeup-roving-tabindex');
+const typeahead = require('makeup-typeahead');
 const eventUtils = require('../../common/event-utils');
 const menuUtils = require('../../common/menu-utils');
-const NodeListUtils = require('../../common/nodelist-utils');
```
Comment (agliga) [inline src/components/ebay-menu/component.js:7]: Can we remove this whole nodelist-utils file and it's test? If it's not being used anymore no need to keep it around.

File: src/components/ebay-menu/test/test.browser.js [hunk 10]
```diff
@@ -11,6 +11,56 @@ afterEach(cleanup);
...
+
+        expect(component.getByText(firstItemText).parentElement.getAttribute('tabindex')).to.equal('0');
+    });
+
+    it('restarts the search from the beginning after it waits', async() => {
+        await fireEvent.click(component.getByText(firstItemText));
+        await fireEvent.keyPress(component.getByText(firstItemText), { key: 'a', code: 65 });
+        await fireEvent.keyPress(component.getByText(firstItemText), { key: 'l', code: 76 });
+        await fireEvent.keyPress(component.getByText(firstItemText), { key: 'b', code: 66 });
+        setTimeout(() => {}, 100);
```
Comment (agliga) [inline src/components/ebay-menu/test/test.browser.js:54]: Can we do this instead
`await new Promise(resolve => setTimeout(resolve, 100));`
This seems cleaner and then I don't think you need the `waitFor`. 
It just bugs me seeing the empty function call

File: src/components/ebay-menu/test/test.browser.js [hunk 11]
```diff
@@ -11,6 +11,56 @@ afterEach(cleanup);
...
+    describe('first', () => {
+        beforeEach(async() => {
+            await fireEvent.click(component.getByText(firstItemText));
+            await fireEvent.keyPress(component.getByText(firstItemText), { key: 'a', code: 65 });
+            await fireEvent.keyPress(component.getByText(firstItemText), { key: 'l', code: 76 });
+            await fireEvent.keyPress(component.getByText(firstItemText), { key: 'c', code: 67 });
+        });
+
+        it('shows the correct item in focus when the user types', async() => {
+            await waitFor(() => {
```
Comment (agliga) [inline src/components/ebay-menu/test/test.browser.js:32]: Did you try to do this without the `waitFor`, not sure if it's needed. 
In the text below you have a `waitFor` and in there there are other `fireEvents` and an assertion without a `waitFor`

**Review by agliga — COMMENTED** [review]
Couple of final nitpicks, after that it should be good.

One more request, can we add a little blurb about this functionality in the `README.md`, so that teams will know it exists.

**Review by agliga — APPROVED** [review]
